lib/Expression_gen.ml

```diff
@@ -1,262 +1,313 @@
+    ; ( Divide
+      , fun ppf es ->
+          if
+            is_matrix (second es)
+            && (is_matrix (first es) || is_row_vector (first es))
+          then pp_binary ppf "mdivide_right(%a, %a)" es
+          else pp_scalar_binary ppf "(%a / %a)" "divide(%a, %a)" es )
+    ; (Modulo, fun ppf es -> pp_binary ppf "modulus(%a, %a)" es)
+    ; (LDivide, fun ppf es -> pp_binary ppf "mdivide_left(%a, %a)" es)
+    ; ( EltTimes
+      , fun ppf es ->
+          pp_scalar_binary ppf "(%a * %a)" "elt_multiply(%a, %a)" es )
+    ; ( EltDivide
+      , fun ppf es -> pp_scalar_binary ppf "(%a / %a)" "elt_divide(%a, %a)" es
+      )
+    ; (Pow, fun ppf es -> pp_binary ppf "pow(%a, %a)" es)
+    ; (Equals, fun ppf es -> pp_binary ppf "logical_eq(%a, %a)" es)
+    ; (NEquals, fun ppf es -> pp_binary ppf "logical_neq(%a, %a)" es)
+    ; (Less, fun ppf es -> pp_binary ppf "logical_lt(%a, %a)" es)
+    ; (Leq, fun ppf es -> pp_binary ppf "logical_lte(%a, %a)" es)
+    ; (Greater, fun ppf es -> pp_binary ppf "logical_gt(%a, %a)" es)
+    ; (Geq, fun ppf es -> pp_binary ppf "logical_gte(%a, %a)" es) ]
+    |> List.map ~f:(fun (k, v) -> (Operators.operator_name k, v))
+    |> Map.Poly.of_alist_exn
+  in
+  let misc_special_math_fns =
+    Map.Poly.of_alist_exn
+      [ ("lmultiply", fun ppf es -> pp_binary ppf "multiply_log(%a, %a)" es)
+      ; ( "lchoose"
+        , fun ppf es -> pp_binary ppf "binomial_coefficient_log(%a, %a)" es )
+      ; ("target", fun ppf _ -> pf ppf "get_lp(lp__, lp_accum__)")
+      ; ("get_lp", fun ppf _ -> pf ppf "get_lp(lp__, lp_accum__)")
+      ; ( "max"
+        , fun ppf es ->
+            if List.length es = 2 then pp_binary ppf "std::max(%a, %a)" es
+            else pp_ordinary_fn ppf f es )
+      ; ( "min"
+        , fun ppf es ->
+            if List.length es = 2 then pp_binary ppf "std::min(%a, %a)" es
+            else pp_ordinary_fn ppf f es )
+      ; ( "ceil"
+        , fun ppf es ->
+            if is_scalar (first es) then pp_unary ppf "std::ceil(%a)" es
+            else pp_ordinary_fn ppf f es ) ]
+  in
+  let read_data_or_param ppf es =
+    let i_or_r =
+      match ut with
+      | Ast.UInt -> "i"
+      | UReal -> "r"
+      | UVector | URowVector | UMatrix | UArray _
+       |UFun (_, _)
+       |UMathLibraryFunction ->
+          raise_s [%message "Can't ReadData of " (ut : unsizedtype)]
+    in
+    pf ppf "context__.vals_%s(%a)" i_or_r pp_expr (List.hd_exn es)
+  in
+  let mir_special_functions =
+    Map.Poly.of_alist_exn
+      [ (fn_length, fun ppf es -> pp_unary ppf "length(%a)" es)
+      ; ( fn_make_array
+        , fun ppf es -> pf ppf "{%a}" (list ~sep:comma pp_expr) es )
+      ; (fn_read_data, read_data_or_param)
+      ; (fn_read_param, read_data_or_param)
+      ; (fn_constrain, pp_constrain_funapp "constrain")
+      ; (fn_unconstrain, pp_constrain_funapp "unconstrain")
+      (* XXX Fill in the rest of the functions at the bottom of Mir.ml *)
```

Comment:
Where did these come from?  Why not just generate these like other function?  Is it the target reference that we need?

Reply:
Which ones?

Reply:
It looks like those are special functions added for reading, constraining, and unconstraining parameters.  I'm asking why they're not just treated like other functions that are resolved in the AST to MIR translation.